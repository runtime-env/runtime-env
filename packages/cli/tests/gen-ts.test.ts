import fs from "fs";
import path from "path";
import { spawnSync } from "child_process";
import { tmpdir } from "tmp";

const schemaFile = path.resolve(__dirname, ".runtimeenvschema.json");

describe("integration - gen-ts", () => {
  test("globalVariableName", () => {
    const result = spawnSync(
      "node",
      ["../bin/runtime-env.js", "gen-ts", "--global-variable-name", "env"],
      {
        encoding: "utf8",
        stdio: "pipe",
        cwd: __dirname,
      },
    );
    expect(result.status).toBe(0);
    expect(result.output).toMatchSnapshot();
  });

  test("schemaFile", () => {
    const result = spawnSync(
      "node",
      ["../bin/runtime-env.js", "gen-ts", "--schema-file", schemaFile],
      {
        encoding: "utf8",
        stdio: "pipe",
        cwd: __dirname,
      },
    );
    expect(result.status).toBe(0);
    expect(result.output).toMatchSnapshot();
  });

  test("schemaFile - invalid", () => {
    const result = spawnSync(
      "node",
      ["../bin/runtime-env.js", "gen-ts", "--schema-file", "invalid"],
      {
        encoding: "utf8",
        stdio: "pipe",
        cwd: __dirname,
      },
    );
    expect(result.status).toBe(1);
    expect(result.output).toMatchSnapshot();
  });

  test("outputFile", () => {
    const result = spawnSync(
      "node",
      [
        "../bin/runtime-env.js",
        "gen-ts",
        "--output-file",
        path.resolve(tmpdir, "runtime-env-gen-ts-output-file.ts"),
      ],
      {
        encoding: "utf8",
        stdio: "pipe",
        cwd: __dirname,
      },
    );
    expect(result.status).toBe(0);
    expect(result.output).toMatchSnapshot();
    expect(
      fs.readFileSync(
        path.resolve(tmpdir, "runtime-env-gen-ts-output-file.ts"),
        "utf8",
      ),
    ).toMatchSnapshot();
  });
});
