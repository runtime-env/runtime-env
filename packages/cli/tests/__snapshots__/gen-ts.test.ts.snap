// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`integration - gen-ts schemaFile - invalid 1`] = `
[
  null,
  "",
  "[@runtime-env/cli]
  schema file not found: no such file, open 'invalid'
",
]
`;

exports[`integration - gen-ts schemaFile 1`] = `
[
  null,
  "// Generated by '@runtime-env/cli'

type Primitive = undefined | null | boolean | string | number;

type DeepReadonly<T> = T extends Primitive
  ? T
  : T extends Array<infer U>
    ? DeepReadonlyArray<U>
    : DeepReadonlyObject<T>;

type DeepReadonlyArray<T> = ReadonlyArray<DeepReadonly<T>>;

type DeepReadonlyObject<T> = {
  readonly [K in keyof T]: DeepReadonly<T[K]>;
};

declare global {
  var runtimeEnv: RuntimeEnv
}

export type RuntimeEnv = DeepReadonly<{
  FOO: string;
  BAR?: {
    BAZ?: string;
  };
}>

",
  "",
]
`;

exports[`integration - gen-ts globalVariableName 1`] = `
[
  null,
  "// Generated by '@runtime-env/cli'

type Primitive = undefined | null | boolean | string | number;

type DeepReadonly<T> = T extends Primitive
  ? T
  : T extends Array<infer U>
    ? DeepReadonlyArray<U>
    : DeepReadonlyObject<T>;

type DeepReadonlyArray<T> = ReadonlyArray<DeepReadonly<T>>;

type DeepReadonlyObject<T> = {
  readonly [K in keyof T]: DeepReadonly<T[K]>;
};

declare global {
  var env: Env
}

export type Env = DeepReadonly<{
  FOO: string;
  BAR?: {
    BAZ?: string;
  };
}>

",
  "",
]
`;

exports[`integration - gen-ts outputFile 1`] = `
[
  null,
  "",
  "",
]
`;

exports[`integration - gen-ts outputFile 2`] = `
"// Generated by '@runtime-env/cli'

type Primitive = undefined | null | boolean | string | number;

type DeepReadonly<T> = T extends Primitive
  ? T
  : T extends Array<infer U>
    ? DeepReadonlyArray<U>
    : DeepReadonlyObject<T>;

type DeepReadonlyArray<T> = ReadonlyArray<DeepReadonly<T>>;

type DeepReadonlyObject<T> = {
  readonly [K in keyof T]: DeepReadonly<T[K]>;
};

declare global {
  var runtimeEnv: RuntimeEnv
}

export type RuntimeEnv = DeepReadonly<{
  FOO: string;
  BAR?: {
    BAZ?: string;
  };
}>
"
`;
